File: _cucumber/lib/helpers/it_admin.rb
class ItAdmin

  def initialize
    @email = "rhd-autotest+sid_#{$session_id}_#{Faker::Lorem.characters(5)}@redhat.com"
    @username = Faker::Lorem.characters(10)
    @greeting = 'Mr.'
    @first_name = Faker::Name.first_name
    @last_name = Faker::Name.last_name
    @password = 'P@$$word01'
    @company_name = Faker::Company.name
    @address_line_one = Faker::Address.street_address
    @city = Faker::Address.city_prefix
    @postal_code = Faker::Address.postcode
    @country = 'UK'
    @phone_number = '0191 1111111'
    @full_name = "#{@first_name} #{@last_name}".upcase
    @locale = 'en_GB'
  end

  def create_simple_user
    @account_details = {
        :disabled => false,
        :emailAddress => @email,
        :emailVerified => true,
        :joinDate => (Time.now.to_i).floor
    }
    begin
      RestClient::Request.execute(
          :method => :post,
          :url => 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/simpleuser/suser/thirdparty?system=rhd',
          :payload => @account_details.to_json,
          :headers => {
              :content_type => :json,
              :accept => :json
          },
      )
    rescue => e
      raise("Failed to create a new simple user: Exception was #{e}")
    end

    change_password(@email, @password)

    return {
        :email => @email,
        :username => @email,
        :password => @password,
        :first_name => @first_name,
        :last_name => @last_name,
        :full_name => @full_name,
        :company_name => @company_name,
        :country => 'United Kingdom'
    }
  end

  def change_password(email, password)
    email_address = CGI.escape(email)
    begin
      RestClient::Request.execute(
          :method => :put,
          :url => "http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/simpleuser/pwd/login=#{email_address}?password=#{password}",
          :headers => {
              :content_type => :json,
              :accept => :json
          },
      )
    rescue => e
      raise("Failed to update password: #{e}")
    end
  end

  def create_full_user

    tries = 0

    @account_details = {
        :userType => 'P',
        :login => @username,
        :active => true,
        :password => @password,
        :system => 'WEB',

        :personalInfo => {
            :company => @company_name,
            :firstName => @first_name,
            :title => 'Test Engineer',
            :greeting => @greeting,
            :lastName => @last_name,
            :email => @email,
            :emailConfirmed => true,
            :phoneNumber => @phone_number,
            :password => @password,
            :locale => @locale
        },

        :personalSite => {
            :defaultSite => true,
            :siteType => 'MARKETING',

            :address => {
                :address1 => @address_line_one,
                :countryCode => @country,
                :postalCode => @postal_code,
                :city => @city,
                :poBox => false
            }
        }
    }

    begin
      RestClient::Request.execute(
          :method => :post,
          :url => 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/createwithoutwelcome',
          :payload => @account_details.to_json,
          :verify_ssl => false,
          :headers => {
              :content_type => :json,
              :accept => :json
          },
      )

    rescue => e
      tries += 1
      if tries <= 10
        p "attempt #{tries}"
        retry
      else
        raise("Failed to create a new user: Exception was #{e}")
      end
    end

    p "Successfully created full site user with #{@email}"

    return {
        :email => @email,
        :username => @email.gsub('@redhat.com', '').gsub('+', '-'),
        :password => @password,
        :first_name => @first_name,
        :last_name => @last_name,
        :full_name => @full_name,
        :company_name => @company_name,
        :country => 'United Kingdom',
    }
  end

  def find_user_by_email(email)
    email_address = CGI.escape(email)
    @user_profile = RestClient::Request.execute(
        :method => :get,
        :url => "http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/email=#{email_address}",
        :headers => {
            :content_type => :json,
            :accept => :json
        }
    )
  end

  def disable_user(email)
    user_profile = find_user_by_email(email)
    hash = JSON.parse user_profile

    @account_details = {
        :id => hash[0]['id'],
        :login => @username,
        :userType => 'P',
        :active => 'false',
        :password => @password,
        :system => 'WEB',

        :personalInfo => {
            :company => @company_name,
            :firstName => @first_name,
            :title => 'Test Engineer',
            :greeting => @greeting,
            :lastName => @last_name,
            :email => @email,
            :emailConfirmed => 'true',
            :phoneNumber => @phone_number,
            :password => @password,
            :locale => @locale
        },

        :personalSite => {
            :id => hash[0]['personalSite']['id'],
            :defaultSite => 'true',
            :siteType => 'MARKETING',

            :address => {
                :address1 => @address_line_one,
                :countryCode => @country,
                :postalCode => @postal_code,
                :city => @city,
                :poBox => 'false'
            },
            :active => 'false',
        }
    }
    tries = 0
    begin
      RestClient::Request.execute(
          :method => :post,
          :url => 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/update',
          :payload => @account_details.to_json,
          :verify_ssl => false,
          :headers => {
              :content_type => :json,
              :accept => :json
          },
      )

    rescue => e
      tries += 1
      if tries <= 10
        p "attempt #{tries}"
        retry
      else
        raise("Failed to update user: Exception was #{e}")
      end
    end
  end

end

File: _tests/e2e/keycloak/features/support/rest/IT.admin.js
const request = require('sync-request');
const qs = require('querystring');
const await = require('asyncawait/await');

// We have outdated certs in our staging environment. This allows us to access the staging keycloak admin.
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

/**
 *  IT Admin contains a set of functions that use the admin service to generate and manipulate test data
 *  (NOTE: Used for Staging environment ONLY)
 */
class ItAdmin {


    /**
     * Creates an active OpenShift account
     * @param  user - User credentials
     */
    createSimpleUser(user) {
        let url = 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/simpleuser/suser/thirdparty?system=rhd';
        try {
            await(request('POST', url, {
                headers: {
                    "Content-Type": "application/json",
                },
                json: {
                    "disabled": false,
                    "emailAddress": user['email'],
                    "emailVerified": true,
                }
            }));
        } catch (e) {
            throw new Error(`Failed to create simple user: ${e}`)
        }

        this._changePassword(user['email'], user['password']);
        console.log(`Created simple user with email: ${user['email']}, and password: ${user['password']}`)
    }

    /**
     * Creates an active Customer portal account
     * @param  user - User credentials
     */
    createFullUser(user) {
        let url = 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/createwithoutwelcome';
        let res = await(request('POST', url, {
            headers: {
                "Content-Type": "application/json",
            },
            json: {
                "userType": 'P',
                "login": user['email'],
                "active": true,
                "password": user['password'],
                "system": 'WEB',

                "personalInfo": {
                    "company": user['company'],
                    "firstName": user['firstName'],
                    "title": user['title'],
                    "greeting": user['title'],
                    "lastName": user['lastName'],
                    "email": user['email'],
                    "emailConfirmed": true,
                    "phoneNumber": user['phoneNumber'],
                    "password": user['password'],
                    "locale": 'en_GB'
                },

                "personalSite": {
                    "defaultSite": true,
                    "siteType": 'MARKETING',

                    "address": {
                        "address1": 'Science Central, Bath Ln',
                        "countryCode": 'GB',
                        "postalCode": 'NE4 5TF',
                        "city": 'Newcastle upon Tyne',
                        "poBox": false
                    }
                }
            }
        }));

        if (res.statusCode !== 200) {
            throw new Error('Failed to create full site user')
        }
        console.log(`Created full-site user with email: ${user['email']}, and password: ${user['password']}`)
    }

    /**
     * Changes password for given email address
     * @private
     * @param  email    - email that you wish to change password for
     * @param  password - updated password
     */
    _changePassword(email, password) {
        let encodedEmail = qs.escape(email);
        let url = `http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/simpleuser/pwd/login=${encodedEmail}?password=${password}`;
        try {
            await(request('PUT', url, {
                headers: {"Content-Type": "application/json"},
            }));
        } catch (e) {
            throw new Error(`Failed to update password: ${e}`)

        }
    }

    /**
     * Searches for a user by registered email address
     * @param  email    - email address of the user you wish to search for
     */
    findUserByEmail(email) {
        let encodedEmail = qs.escape(email);
        let url = `http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/email=${encodedEmail}`;
        let res = request('GET', url);
        return JSON.parse(res.getBody().toString('utf8'));
    }

    /**
     * Disable a registered user
     * @param  user - hash of user details
     */
    disableUser(user) {
        let userProfile = this.findUserByEmail(user['email']);
        let url = 'http://servicejava.edge.stage.ext.phx2.redhat.com/svcrest/user/v3/update';
        let res = await(request('POST', url, {
            headers: {
                "Content-Type": "application/json",
            },
            json: {
                "id": userProfile[0]['id'],
                "login": user['email'],
                "userType": 'P',
                "active": false,
                "password": user['password'],
                "system": 'WEB',

                "personalInfo": {
                    "company": user['company'],
                    "firstName": user['firstName'],
                    "title": user['title'],
                    "greeting": user['title'],
                    "lastName": user['lastName'],
                    "email": user['email'],
                    "emailConfirmed": 'true',
                    "phoneNumber": user['phoneNumber'],
                    "password": user['password'],
                    "locale": 'en_GB'
                },

                "personalSite": {
                    "id": userProfile[0]['personalSite']['id'],
                    "defaultSite": true,
                    "siteType": 'MARKETING',

                    "address": {
                        "address1": 'Science Central, Bath Ln',
                        "countryCode": 'GB',
                        "postalCode": 'NE4 5TF',
                        "city": 'Newcastle upon Tyne',
                        "poBox": false
                    },
                    "active": false
                }
            }
        }));

        console.log(JSON.parse(res.getBody().toString('utf8')));

        if (res.statusCode === 500) {
            throw new Error('Failed to disable user account')
        }
    }

}

module.exports = ItAdmin;

